// @tanstack/react-form/build/lib/useField.legacy.cjs
'use strict';

var React = require('react');
var reactStore = require('@tanstack/react-store');
var formCore = require('@tanstack/form-core');
var formContext = require('./formContext.legacy.cjs');
var useIsomorphicLayoutEffect = require('./utils/useIsomorphicLayoutEffect.legacy.cjs');

function useField(opts) {
  // Get the form API either manually or from context
  const {
    formApi,
    parentFieldName
  } = formContext.useFormContext();
  const [fieldApi] = React.useState(() => {
    const name = (typeof opts.index === 'number' ? [parentFieldName, opts.index, opts.name] : [parentFieldName, opts.name]).filter(d => d !== undefined).join('.');
    const api = new formCore.FieldApi({
      ...opts,
      form: formApi,
      name: name
    });
    api.Field = Field;
    return api;
  });

  /**
   * fieldApi.update should not have any side effects. Think of it like a `useRef`
   * that we need to keep updated every render with the most up-to-date information.
   */
  useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(() => {
    fieldApi.update({
      ...opts,
      form: formApi
    });
  });
  reactStore.useStore(fieldApi.store, opts.mode === 'array' ? state => {
    return [state.meta, Object.keys(state.value || []).length];
  } : undefined);

  // Instantiates field meta and removes it when unrendered
  useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(() => fieldApi.mount(), [fieldApi]);
  return fieldApi;
}

// export type FieldValue<TFormData, TField> = TFormData extends any[]
//   ? TField extends `[${infer TIndex extends number | 'i'}].${infer TRest}`
//     ? DeepValue<TFormData[TIndex extends 'i' ? number : TIndex], TRest>
//     : TField extends `[${infer TIndex extends number | 'i'}]`
//     ? TFormData[TIndex extends 'i' ? number : TIndex]
//     : never
//   : TField extends `${infer TPrefix}[${infer TIndex extends
//       | number
//       | 'i'}].${infer TRest}`
//   ? DeepValue<
//       DeepValue<TFormData, TPrefix>[TIndex extends 'i' ? number : TIndex],
//       TRest
//     >
//   : TField extends `${infer TPrefix}[${infer TIndex extends number | 'i'}]`
//   ? DeepValue<TFormData, TPrefix>[TIndex extends 'i' ? number : TIndex]
//   : DeepValue<TFormData, TField>
// type Test1 = FieldValue<{ foo: { bar: string }[] }, 'foo'>
// //   ^?
// type Test2 = FieldValue<{ foo: { bar: string }[] }, 'foo[i]'>
// //   ^?
// type Test3 = FieldValue<{ foo: { bar: string }[] }, 'foo[2].bar'>
// //   ^?
function Field({
  children,
  ...fieldOptions
}) {
  const fieldApi = useField(fieldOptions);
  return /*#__PURE__*/React.createElement(formContext.formContext.Provider, {
    value: {
      formApi: fieldApi.form,
      parentFieldName: fieldApi.name
    },
    children: formCore.functionalUpdate(children, fieldApi)
  });
}

exports.Field = Field;
exports.useField = useField;
